oitg/fitting/attenuated_sinusoid.py

```diff
@@ -0,0 +1,194 @@
+"""Fit an attenuated sinusoid with optional dead-time to data"""
+import numpy as np
+from scipy.signal import lombscargle
+from .FitBase import FitBase
+
+
+def fitting_function(t, p_dict):
+    """attenuated sinusoid with dead time"""
+    y = np.exp(-p_dict['rate'] * t) * \
+        (p_dict['a'] * np.sin((t - p_dict['t_dead']) * p_dict['omega']
+                              + p_dict['phi'])
+         + p_dict['c_offset']) + p_dict['c_equ']
+    # hold constant during dead time
+    return np.where(t > np.min(t) + p_dict['t_dead'], y,
+                    p_dict['c_offset'] + p_dict['c_equ'] +
+                    p_dict['a'] * np.sin(p_dict['phi'])
+                    )
+
+
+def init_all(t, y, p_dict):
+    """Initialise parameters under the general assumptions that:
+    t & y are 1D arrays of real numbers
+    associated data coordinates share the same index in t & y
+
+    Note: Data may be in any order and arbitrarily spaced"""
+    # p_dict is an instance of FitParameters
+    # user-inintialised values are not overwritten
+    # non user-initialised parameters are pre-initialised to 0
+
+    # sort data by time (x-axis)
+    mask = np.argsort(t)
+    t = t[mask]
+    y = y[mask]
+
+    p_dict['t_dead'] = 0.0
+    p_dict['rate'] = 0.0
+
+    min_step = np.min(t[1:] - t[:-1])
+    duration = t[-1] - t[0]
+    f_max = 0.5 / min_step
+    f_min = 0.5 / duration
+
+    omega_list = 2 * np.pi * np.linspace(f_min, f_max, int(f_max / f_min))
+    if ('c_equ' in p_dict.constant_parameter_names or
+        'c_equ' in p_dict.initialised_parameter_names):
+        temp = (y - p_dict['c_equ']) * np.exp(p_dict['rate'] * t)
+        pgram = lombscargle(t, temp, omega_list, precenter=True)
+    else:
+        pgram = lombscargle(t, y, omega_list, precenter=True)
+    p_dict['omega'] = omega_list[np.argmax(pgram)]
+
+    temp = pgram[np.searchsorted(omega_list, p_dict['omega'])]
+    p_dict['a'] = np.sqrt(temp * 4 / len(y))
+
+    # having estimated 'omega', average over a period
+    t_period = 2 * np.pi / p_dict['omega']
+    p_dict['c_equ'] = np.mean(y[np.argwhere(t > t[-1] - 2 * t_period)])
+    p_dict['c_offset'] = np.mean(y[np.argwhere(t < t[0] + t_period)]) - \
+                         p_dict['c_equ']
+
+    y0_norm = (y[0] - p_dict['c_equ'] - p_dict['c_offset']) / p_dict['a']
+    if y0_norm <= -1.0:
+        p_dict['phi'] = -np.pi / 2
+    elif y0_norm >= 1.0:
+        p_dict['phi'] = np.pi / 2
+    else:
+        p_dict['phi'] = np.arcsin(y0_norm)
+    return p_dict  # input is changed due to mutability of dict!
+
+
+def derived_params(p_dict, p_error_dict):
+    """calculate commonly used derived parameters and their error
+
+    This method neglects parameter covariances!"""
+    p_dict['t_pi'] = p_dict['t_dead'] + np.pi / p_dict['omega']
```

Comment:
NB: This isn't actually the point of maximal population transfer for finite decay times, see rabi_flop. (There is probably an analytical solution, though.)

Reply:
I've updated it to the analytical result.